pkg/tools/etcd_helper_watch.go

```diff
@@ -274,6 +281,9 @@ func (w *etcdWatcher) decodeObject(node *etcd.Node) (runtime.Object, error) {
 		}
 	}
 
+	if node.ModifiedIndex != 0 {
+		w.cache.addToCache(node.ModifiedIndex, obj)
```

Comment:
Maybe the race is because someone later modifies obj? Maybe you need to deepcopy before returning.